jobs/integration/test_opa.py

```diff
@@ -0,0 +1,419 @@
+            self.tools.controller_name,
+            self.tools.connection,
+        )
+        log("Delete example policy")
+        await kubectl(self.model, f"delete -f {remote_path}")
+
+    @pytest.fixture(scope="module")
+    async def storage_pool(self, model, k8s_model, tools):
+        # This assumes that we are running on vsphere, should it be
+        # extended to run on other clouds as well?
+        control_plane_app = model.applications["kubernetes-control-plane"]
+        control_plane_unit = control_plane_app.units[0]
+        remote_path = "/tmp/storage-class.yaml"
+        storage_pool_name = "teststorage"
+        await scp_to(
+            templates / "integrator-charm-data" / "vsphere" / "storage-class.yaml",
+            control_plane_unit,
+            remote_path,
+            tools.controller_name,
+            tools.connection,
+        )
+        log("Creating storage class")
+        try:
+            await kubectl(model, f"create -f {remote_path}")
+            await tools.run(
+                "juju",
+                "create-storage-pool",
+                storage_pool_name,
+                "kubernetes",
+                "-m",
+                tools.k8s_connection,
+                "storage-class=mystorage",
+            )
+            yield storage_pool_name
+        finally:
+            log("Removing storage class")
+            try:
+                await kubectl(model, f"delete -f {remote_path}")
+            finally:
+                await tools.run(
+                    "juju",
+                    "remove-storage-pool",
+                    "-m",
+                    tools.k8s_connection,
+                    storage_pool_name,
+                )
+
+    async def _validate_audit_actions(self, unit):
+        log("Running list-violations action")
+        total_violations = 0
+        while total_violations == 0:
+            action = await unit.run_action("list-violations")
+            await action.wait()
+            assert action.status == "completed"
+            assert "constraint-violations" in action.results
+            constraint_violations = json.loads(action.results["constraint-violations"])
+            if constraint_violations:
+                total_violations = constraint_violations[0].get("total-violations")
+            await asyncio.sleep(5)
+
+        log("Running get-violation action")
+        action = await unit.run_action(
+            "get-violation",
+            **{
+                "constraint-template": "K8sRequiredLabels",
+                "constraint": "ns-must-have-gk",
+            },
+        )
+        await action.wait()
+        assert len(json.loads(action.results["violations"])) > 0
+
+    async def test_opa_webhook(self, storage_pool):
+        log("Deploying the gatekeeper charm")
+        webhook = await self.k8s_model.deploy(
+            "gatekeeper-controller-manager",
+            channel=self.tools.charm_channel,
+            trust=True,
+        )
```

Comment:
The fixture would be used only for one test, I don't think it would offer much.

Reply:
fixtures do aid in guaranteed teardown which is a bit nice than

```python
   try:
     # setup stuff
     # run test 
   finally:
     # teardown
```

i think @mateoflorido is on to something when he's advocating for a fixture here.  You can get a test ERROR because of a setup/teardown failure and test FAILED when the validation fails.

i'm not a fan of the style of nested try blocks, but it seems like it will work, but this test method is really way too big.  It's not testing one thing -- but everything. 


why not `test_opa_charm_ready`, `test_validate_create_ns_no_policy`, `test_validate_create_ns_policy_enforcement` ... etc?

then that fixture becomes a `module` based fixture that runs one, is used by each small test, and then cleans up when all the tests in the module are run.